### ISSUE TITLE
Exporting large number of documents cause memory problem

### ISSUE THREAD

**Issue by bushev** [OpenKitten/MongoKitten#122]
Hello!

I'm trying to export a large number of documents. But getting memory exhausted error. What I'm doing wrong? Do I need to clear something on each iteration?

```swift
func testExport() {
    
    let server = try! Server("mongodb://176.112.204.83:27017")
    
    let database = server["test-db"]
    let collection = database["images"]
    
    let documents: MongoKitten.CollectionSlice<BSON.Document> = try! collection.find()
    
    while let document = documents.next() {
        
        print("DOC: " + document.makeExtendedJSONString(typeSafe: true))
    }
    
    print("Done")
}
```

**Comment by Joannis** [comment]
What is the size of a single document? Does it contain Binary data? These will factor in when converting to JSON.

**Comment by bushev** [comment]
Hi, @Joannis!

Document size (avgObjSize) is 775 Bytes. It doesn't contain Binary data.

Document example:
![image](https://user-images.githubusercontent.com/7535766/27004660-8d33953a-4e15-11e7-87ff-e356c9fc44d2.png)

I have about ~224K of such documents. Application memory usage is about ~2GB before a crash (OOM).

**Comment by Joannis** [comment]
Oh, that's bad. Is there any way you can reproduce this scenario? Or see what is consuming the memory?

**Comment by bushev** [comment]
Hello @Joannis!

Looks like memory leaks in `makeExtendedJSON(typeSafe: true).serializedString()`.

![image](https://user-images.githubusercontent.com/7535766/27049096-eee6ba76-4fb6-11e7-9241-ea4d5178a32c.png)

I see that app allocates a lot of memory with this call stack:

![image](https://user-images.githubusercontent.com/7535766/27049211-3c8a6afc-4fb7-11e7-8619-35449635fb02.png)

**Comment by Joannis** [comment]
Ah :| Thanks so much! I'll be onto this now

**Comment by Joannis** [comment]
Could you try this with the latest BSON version? I think there was a memory leak there.

**Comment by bushev** [comment]
Hello @Joannis,

it's still leaks:

![image](https://user-images.githubusercontent.com/7535766/27150500-23ff02dc-5150-11e7-8ee7-cd0b18536810.png)

Used memory is released only after the completion of the loop over documents.

![image](https://user-images.githubusercontent.com/7535766/27151468-52428774-5153-11e7-8cef-d8a433a1c1d2.png)

You can reproduce this scenario using the following code (using large images collection on my test server):

```swift
func testExport() {
    
    let server = try! Server("mongodb://176.112.204.83:27017")
    
    let database = server["test-db"]
    let collection = database["images"]
    
    let documents: MongoKitten.CollectionSlice<BSON.Document> = try! collection.find()

    for document in documents {

        print("DOC: " + document.makeExtendedJSON(typeSafe: true).serializedString())
    }
    
    print("Done")
}
```

**Comment by Joannis** [comment]
That makes sense! Images (binary) will be converted to Base64 in ExtendedJSON. That takes an insane amount of memory probably.

**Comment by bushev** [comment]
There are no binary fields in the image collection, strings and numbers only.

**Comment by Joannis** [comment]
Hah. I'm still diving in anyways. Even with Base64 the RAM usage would be insane. This isn't making it any better. I don't know how it can rise 600MB even.

**Comment by Joannis** [comment]
I found _some_ improvement points. But I'm still struggling to find out the exact cause.

**Comment by Joannis** [comment]
There were memory leaks, so your memory leak checks are correct. But the huge amount of memory being used isn't leaking. It's the conversion from BSON->JSON that's slow. Conversion can't become any faster, but I can write a specialized serialization algorithm.

**Comment by Joannis** [comment]
The latest version on master has a new API for extendedJSON. It's not tagged yet, we're going to run this version of MongoKitten on production systems before tagging this. It's working, but MongoKitten has had 2 of the lowest internal functions reworked, so taking extra care is important.

It's tested to be fine, so if you want you can run `swift package edit MongoKitten --revision mongokitten4` and it'll make your project use the latest version of MongoKitten 4 (non-tagged). From here you'll experience both an improvement to all operations as well as extremely low memory-usage and high-performance extendedJSON serialization. Deserialization is still the same performance and memory usage, but I don't think there's a big issue in that.

The API you're looking for is this: https://github.com/OpenKitten/MongoKitten/blob/mongokitten4/Sources/ExtendedJSON/ExtendedJSON%2BEfficientSerialization.swift#L55

**Comment by bushev** [comment]
Great, I will test new extendedJSON soon. Thanks!

**Comment by bushev** [comment]
Sad news, unfortunately, new API was not solved memory problem. But also application doesn't release all memory after loop finished...

I used new extendedJSON function:

```swift
document.makeExtendedJSONString(typeSafe: true)
```

![image](https://user-images.githubusercontent.com/7535766/27252558-172a5554-536a-11e7-9d01-4bca58bc8ed5.png)

**Comment by Joannis** [comment]
Oh man.. is this a new issue? The memory not being released?

I don't know what could possibly be causing this.
You're still running this with the `testExport` function described above, right?

**Comment by Joannis** [comment]
I can't even start to imagine this happening. Locally it reduced the memory usage by 95%.
I feel like there's something off about your documents or code that I can't foresee.

**Comment by bushev** [comment]
My test function (I updated descriptino as well):

![image](https://user-images.githubusercontent.com/7535766/27252716-c722b93a-536d-11e7-8839-4d73d67bc2cd.png)

Memory is not released after "Done" is printed:

![image](https://user-images.githubusercontent.com/7535766/27252711-b5a58070-536d-11e7-828c-766535b1da32.png)

Did you able to reproduce it with my `testExport` function?

**Comment by Joannis** [comment]
I was able to reproduce high memory usage and tiny memory leaks. Nothing significant. If you want to help me accelerate the solution you could join my slack https://slackpass.io/openkitten

**Comment by bushev** [comment]
I can confirm that my issue is resolved.

Fixed in: https://github.com/OpenKitten/MongoKitten/commit/1973a4751d0197538251b9353c618a0edfbfa737
